### PR TITLE
Make trailing capture group optional in ImplementationSpecific paths

### PR DESCRIPTION
**Pull request by mmatur**
### What does this PR do?

Introduces `makeTrailingGroupOptional`, which rewrites an `ImplementationSpecific` path like `/original/(.*)` to `/original(?:/(.*))? ` so that the bare path `/original` also matches. The transformation applies to both the router `PathRegexp` rule and the `RewriteTarget` regex. Paths whose first character is a capture group are left unchanged.

### Motivation

The nginx ingress controller handles the trailing-slash edge case implicitly. Without this fix, a request to `/original` returned 404 because the generated `PathRegexp` required a literal `/` after the prefix.

### More

- [x] Added/updated tests
- [ ] Added/updated documentation

### Additional Notes

### CODE REVIEW

File: pkg/provider/kubernetes/ingress-nginx/translator.go [hunk 1]
```diff
@@ -620,3 +623,18 @@ func buildPrefixRule(path string) string {
 	path = strings.TrimSuffix(path, "/")
 	return fmt.Sprintf("(Path(%q) || PathPrefix(%q))", path, path+"/")
 }
+
+// makeTrailingGroupOptional converts an ImplementationSpecific regex path like /foo/(.*)
+// into /foo(?:/(.*))? so it also matches the bare /foo (without trailing slash).
+// Only applies when the path starts with a literal prefix (not a capture group),
+// i.e. when the path does not begin with "/(".
+func makeTrailingGroupOptional(path string) string {
```
Comment (sdelicata) [inline pkg/provider/kubernetes/ingress-nginx/translator.go:669]: Could you add a small table-driven test for `makeTrailingGroupOptional` directly? A few edge cases would be nice to lock down: `/(.*)` left untouched, multiple groups like `/foo/(bar)/(.*)`, and whatever we decide for the `/originalfoo` case above.

File: pkg/provider/kubernetes/ingress-nginx/middleware.go [hunk 2]
```diff
@@ -261,6 +262,9 @@ func (p *Provider) buildRewriteTarget(loc *location) {
 	}
 
 	regex := loc.Path
+	if ptr.Deref(loc.PathType, netv1.PathTypePrefix) == netv1.PathTypeImplementationSpecific {
+		regex = makeTrailingGroupOptional(loc.Path)
```
Comment (sdelicata) [inline pkg/provider/kubernetes/ingress-nginx/middleware.go:276]: Maybe I'm missing something, but I think this reopens the suffix leak from #12931 for a request like `/originalfoo` (the prefix immediately followed by non-slash chars).

The router rule `^/original(?:/(.*))?` matches `/originalfoo`, so it gets routed to the rewrite middleware, and there `ReplaceAllString` only matches `/original` (the optional group fails on the `f`) and leaks the trailing `foo`, so we redirect to `https://bar.example.org/foo`. nginx routes that same request to its implicit `^/` catch-all and redirects to `https://bar.example.org/`. And before this PR `/originalfoo` was a 404, so it's a behavior change too.

The #12931 guard that appends `.*` to consume the whole path lives in `rewritetarget.New` (the `if re.NumSubexp() == 0 && absoluteURLRedirect` branch in `pkg/middlewares/ingressnginx/rewritetarget/rewrite_target.go`), but it doesn't fire here: `makeTrailingGroupOptional` always emits a capturing group, so `NumSubexp()` is 1, not 0.

Relaxing that condition to also append `.*` in the absolute-URL case when there are capture groups fixes it, and the captures stay intact:
`/originalfoo` -> `.../`, `/original/foo` -> `.../foo`, `/original` -> `.../`.
WDYT?

**Review by sdelicata — CHANGES_REQUESTED** [review]

File: pkg/provider/kubernetes/ingress-nginx/middleware.go (hunk 2 above)
Comment (mmatur) [reply pkg/provider/kubernetes/ingress-nginx/middleware.go:276]: You are right, I am moving with this approach. By the way I have added some test on [ingress-nginx-migration](https://github.com/traefik/ingress-nginx-migration/pull/33)

File: pkg/provider/kubernetes/ingress-nginx/translator.go [hunk 3]
```diff
@@ -620,3 +623,36 @@ func buildPrefixRule(path string) string {
...
+			if depth == 0 && i != len(path)-1 {
+				return path
+			}
+		}
+	}
+	if depth != 0 {
+		return path
+	}
+
+	return path[:idx] + "(?:" + path[idx:] + ")?"
```
Comment (sdelicata) [inline pkg/provider/kubernetes/ingress-nginx/translator.go:694]: Maybe I'm wrong, but I think `/originalfoo` still diverges for the non-absolute rewrite case (the one the new `ingress-with-rewrite-target-trailing-group.yml` fixture generates).

`^/original(?:/(.*))?` matches `/originalfoo`, and in the middleware `ReplaceAllString` only matches `/original`, so `/newpath/$1` gives `/newpath/` plus the leftover `foo`, and we proxy `/newpath/foo` upstream. The `.*` guard fixed this for absolute URLs only. nginx wouldn't match `/originalfoo` at all with `location ~* "^/original/(.*)"`, and before this PR it was a 404, so it's a behavior change too.

WDYT about emitting `/original(?:$|/(.*))` instead? It's the pattern ingress-nginx documents itself in https://kubernetes.github.io/ingress-nginx/examples/rewrite/ (`/something(/|$)(.*)`), the capture indexes stay the same, `/original` and `/original/foo` still match, and `/originalfoo` doesn't.

**Review by sdelicata — CHANGES_REQUESTED** [review]

File: pkg/provider/kubernetes/ingress-nginx/translator.go (hunk 3 above)
Comment (mmatur) [reply pkg/provider/kubernetes/ingress-nginx/translator.go:694]: Good catch, the non-absolute case was indeed still leaking.

I tried your pattern first, but the `ingress-nginx-migration` e2e suite caught a regression on the absolute-URL case: for `/originalfoo`, nginx answers `302 https://bar.example.org/` (the request falls through to the implicit `^/` catch-all, whose rewrite still issues the redirect), while `/original(?:$|/(.*))` makes Traefik answer 404.

It turns out nginx behaves differently depending on the rewrite-target:

- absolute URL: unmatched paths hit the catch-all and still get the 302 redirect
- path rewrite: unmatched paths hit the catch-all and get a 404 from the default backend

So instead of changing the pattern, I now only apply `makeTrailingGroupOptional` when the rewrite-target is an absolute URL. For the non-absolute case the regex path is kept as-is, so `/original` and `/originalfoo` return 404 exactly like nginx, and the suffix leak is gone without touching the middleware.

With this, the whole `TestRewriteTargetSuite` [e2e suite](https://github.com/traefik/ingress-nginx-migration/pull/33) passes against a real ingress-nginx, including the `/originalfoo` cases in both modes.

**Review by sdelicata — APPROVED** [review]
Thanks for the changes!
LGTM

**Review by juliens — APPROVED** [review]
